Reftables support

pks-gitlab opened the pull request:
This pull request introduces support for the reftable backend. It depends on the following list of pull requests:

* https://github.com/libgit2/libgit2/pull/7102 introduces the refFormat extension.
* https://github.com/libgit2/libgit2/pull/7111 fixes our handling of pseudo and root refs to align with modern Git.
* https://github.com/libgit2/libgit2/pull/7113 adjusts our tests to not access references via the filesystem directly anymore.
* https://github.com/libgit2/libgit2/pull/7114 is a semi-related fix, but our CI is currently broken due to changes in the MSVC compiler version.

Other than that, all tests are passing and there aren't any memory leaks. But some polishing is still required for non-Linux platforms.

This pull request supersedes https://github.com/libgit2/libgit2/pull/5462.

pks-t committed:
deps: import reftable library

Import the reftable library from commit 4fee6ff3b2 (Merge branch
'ps/reftable-portability', 2026-04-08). This is an exact copy of the
reftable library.

The library will be wired into libgit2 over the next couple of commits.

pks-t committed:
refdb: implement reftable backend

Implement the reftable backend that is used to read and write reftables.
The backend is not yet used anywhere after this commit.

pks-t committed:
deps/reftable: wire up library with CMake

Wire up the reftable library with CMake. At the current point in time
the library is not yet plugged into libgit2 itself.

pks-t committed:
refdb: wire up "reftable" storage format

Wire up the "reftable" storage format via the newly implemented reftable
backend.

pks-t committed:
tests: generate test resources for reftables

Generate test resources for reftables. These resources are basically the
Git repositories we already have, but converted to use the "reftable"
format. For most of the part, this conversion is done by executing `git
refs migrate`.

A couple notes:

  - This require a recent Git upstream version with not-yet-upstreamed
    patches due to a bug in `git refs migrate` with reflogs.

  - The migration command does not yet support repositories with
    worktrees. Those were converted by first removing the worktrees,
    migrating the refs and then recreating them.

  - The HEAD_TRACKER reference in testrepo.git is not recognized as a
    root ref and is thus not automatically migrated.

  - testrepo.git has an empty reflog for refs/heads/with-empty-log that
    does not get migrated.

pks-t committed:
tests: wire up reftable tests

Wire up reftable tests so that they can be executed by setting the
`CLAR_REF_FORMAT` environment variable. This only catches tests that use
`cl_git_sandbox_init()`, but that should cover most of our tests. So
this infrastructure isn't perfect, but for now it's good enough. We may
want to iterate on it in the future.

csware commented:
Any news on this PR?
@ethomson

pks-gitlab commented:
> Any news on this PR? @ethomson

@csware I guess the single biggest thing that these PRs need is review. If  you have the time to help review I think that would help quite a bit to move this forward.

Other than that I also plan to have some of the ICs in my team at GitLab review those PRs in the next Git release cycle.

csware commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+		goto out;
+	}
+
+	*exists = (error == 0);
```
Is the out parameter needed, the error is also returned.

pks-gitlab replied:
@csware Yes, it's part of the `git_refdb_backend` callback interface.

csware commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+
+		entry = git__calloc(1, sizeof(*entry));
+		GIT_ERROR_CHECK_ALLOC(entry);
+		entry->committer = signature;
```
Possible leak of signature in case of error?!

pks-gitlab replied:
True, there were two exit paths where I didn't free the signature. Fixed now.

pks-gitlab commented:
Rebased on latest `main` now that another prerequisite has been merged. Only one more dependency to go!

pks-t committed:
github: wire up reftable tests

Wire up three new CI jobs that exercise libgit2 with the reftable
backend.

pks-gitlab commented:
The last dependency has been merged now with the test refactorings, so I've rebased this pull request again. Note though that this still depends on a couple of changes to the reftable library in upstream Git, which I've sent in for review today via https://lore.kernel.org/git/20260331-pks-reftable-portability-fixes-v1-0-46bfae55c68c@pks.im/.

I think overall this can already be reviewed, as the bulk of the changes are in the reftable backend itself. But I expect some changes in the reftable library based on review feedback.

pks-t committed:
deps/reftable: implement system compatibility code

While the reftable library is mostly decoupled from the Git codebase,
some specific functionality intentionally taps into Git subsystems. To
make porting of the reftable library easier all of these are contained
in "system.h" and "system.c".

Reimplement those compatibility shims so that they work for libgit2.

pks-t committed:
refs: provide `git_reference__is_per_worktree_ref()`

Extract the function `is_per_worktree_ref()` from the "files" backend
and expose it via "refs.h". This function will be reused by the
"reftable" backend.

pks-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+	return error;
+}
+
+static bool is_per_worktree_ref(const char *ref_name)
```
@csware Yup, makes sense.

pks-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+	GIT_ERROR_CHECK_ALLOC(backend);
+
+	if ((error = git_refdb_init_backend(&backend->parent, GIT_REFDB_BACKEND_VERSION)) < 0)
+		goto out;
```
Yup, fixed now.

pks-t committed:
util/git2_util.h: expose `GIT_INLINE_KEYWORD` macro

In our util headers we expose the `GIT_INLINE()` macro to declare a
function as inline, but we don't provide a way to access the keyword
directly.

Expose a new `GIT_INLINE_KEYWORD` to fill this gap. This define will be
used in a subsequent commit.

pks-gitlab commented:
The patch series that makes the reftable library more portable has been merged upstream, so I've updated the pull request with the exact version that we have in upstream Git now.

knayak-gitlab commented on deps/reftable/reftable-system.h:
```diff
@@ -8,11 +8,26 @@
+
+#define poll(fds, fds_len, timeout) p_poll(fds, fds_len, timeout)
+
+#define REFTABLE_ALLOW_BANNED_ALLOCATORS
```
I couldn't understand what this header is for?

pks-gitlab replied:
@knayak-gitlab Do you mean the header specifically, or this specific define here?

knayak-gitlab replied:
This particular define. Nvm though I didn't check the other files. All good here.

knayak-gitlab commented on deps/reftable/reftable-system.h:
```diff
@@ -8,11 +8,26 @@
+
+#define REFTABLE_ALLOW_BANNED_ALLOCATORS
+
+#define inline GIT_INLINE_KEYWORD
```
Okay so the earlier commit makes sense now, we want to specify the inline keyword.

knayak-gitlab commented on deps/reftable/system.c:
```diff
@@ -1,165 +1,259 @@
 uint32_t reftable_rand(void)
 {
-	return git_rand(CSPRNG_BYTES_INSECURE);
+	return rand();
```
Shouldn't we use `git_rand_next()` here?

knayak-gitlab replied:
Is it because of the type? but that shouldn't matter for when converting uint64_t to uint32_t for a random number no?

pks-gitlab replied:
@knayak-gitlab It's mostly that I missed `git_rand_next()`. Will adapt.

knayak-gitlab commented on deps/reftable/system.c:
```diff
@@ -1,165 +1,259 @@
-	if (!tempfile)
-		return REFTABLE_IO_ERROR;
+	if (git__suffixcmp(pattern, ".XXXXXX"))
+		return REFTABLE_API_ERROR;
```
Okay so we only expect users of this API to provide the ".XXXXXX" suffix. In Git this is handled by `git_mkdstemps_mode()`, so here we manually do it. Makes sense.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+
+	*out = NULL;
+
+#ifdef GIT_EXPERIMENTAL_SHA256
```
So sha256 support itself is experimental in libgit2?

pks-gitlab replied:
@knayak-gitlab Yes, exactly.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+	options.hash_id = REFTABLE_HASH_SHA1;
+#endif
+	options.default_permissions = 0666;
+	options.disable_auto_compact = 0;
```
Aren't these two the default anyways?

pks-gitlab replied:
@knayak-gitlab Indeed, will adapt.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+		GIT_ERROR_CHECK_ALLOC(stack);
+		stack->which = which;
+
+		if ((error = git_str_joinpath(&path, parent_directory, "reftable")) < 0 ||
+		    (error = reftable_new_stack(&stack->stack, path.ptr, &options)) < 0) {
```
Shouldn't the error be transformed via `refdb_reftable_error`?

pks-gitlab replied:
Yup, indeed.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+				goto out;
+			}
+
+			write_head = (error > 0);
```
Okay we only want to write head if it doesn't exist when `GIT_REFDB_BACKEND_INIT_FORCE_HEAD` is unset. Makes sense.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+		goto out;
+	}
+
+	if ((error = refdb_reftable_reference_from_record(out, &ref, backend->repo->oid_type)) < 0)
```
This was a bit confusing, since both the output and the goto are named `out`.

pks-gitlab replied:
Will rename to `out_ref`.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+	 * Verify that the current state of the refname matches the expected
+	 * state for non-racy updates.
+	 */
+	if ((error = refdb_reftable_check_ref(data->stack, data->ref->name, data->expected_oid, data->expected_target)) < 0) {
+		data->error = error;
+		git_error_set(GIT_ERROR_REFERENCE, "old reference value does not match");
+		goto out;
+	}
+
+	if ((error = refdb_reftable_check_refname_available(data->stack, NULL, data->ref->name, data->force)) < 0) {
+		data->error = error;
+		goto out;
+	}
```
Both these function independently read the ref from the stack, wouldn't it be better if we supplied it?

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+
+	/* We do not (yet?) support renames across different worktree stacks. */
+	if (git_repository_is_worktree(backend->repo) &&
+	    git_reference__is_per_worktree_ref(old_name) != git_reference__is_per_worktree_ref(new_name)) {
```
But this doesn't validate against two different worktrees, it only ensures both the names are worktree refs.

pks-gitlab replied:
@knayak-gitlab Hm, right. As far as I can see this can only be the case when we have two named worktree refs (`refs/worktree/`) that name different worktrees. So something like this should address this, right?

```
diff --git a/src/libgit2/refdb_reftable.c b/src/libgit2/refdb_reftable.c
index 8b0219a38..e9013f739 100644
--- a/src/libgit2/refdb_reftable.c
+++ b/src/libgit2/refdb_reftable.c
@@ -1311,13 +1311,46 @@ static int refdb_reftable_rename(git_reference **out,
 	data.out = out;
 	data.error = 0;
 
-	/* We do not (yet?) support renames across different worktree stacks. */
+	/*
+	 * Are we in a worktree, but the references aren't both per-worktree
+	 * refs? In that case they cannot be part of the same worktree, but we
+	 * don't support cross-worktree renames yet.
+	 */
 	if (git_repository_is_worktree(backend->repo) &&
 	    git_reference__is_per_worktree_ref(old_name) != git_reference__is_per_worktree_ref(new_name)) {
 		error = GIT_EINVALID;
 		goto out;
 	}
 
+	/*
+	 * Otherwise, are both refs named worktree refs, but for different
+	 * worktrees? If so, we need to bail out, as well.
+	 */
+	if (!git__prefixcmp(old_name, "refs/worktree/") &&
+	    !git__prefixcmp(new_name, "refs/worktree/")) {
+		const char *old_wtname_end = strchr(old_name + strlen("refs/worktree/"), '/');
+		const char *new_wtname_end = strchr(new_name + strlen("refs/worktree/"), '/');
+		size_t old_prefix_len, new_prefix_len;
+
+		if (!old_wtname_end || !new_wtname_end) {
+			error = GIT_EINVALID;
+			goto out;
+		}
+
+		old_prefix_len = p_strnlen(old_name, old_wtname_end - old_name);
+		new_prefix_len = p_strnlen(new_name, new_wtname_end - new_name);
+
+		if (old_prefix_len != new_prefix_len) {
+			error = GIT_EINVALID;
+			goto out;
+		}
+
+		if (git__strncmp(old_name, new_name, old_prefix_len)) {
+			error = GIT_EINVALID;
+			goto out;
+		}
+	}
+
 	if ((error = refdb_reftable_stack_for_refname(&data.stack, backend, old_name)) < 0)
 		goto out;
 
```

knayak-gitlab replied:
Yeah this looks good!

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+			creation = old_log;
+			git__free(creation.refname);
+			creation.refname = git__strdup(new_name);
+			memset(&old_log, 0, sizeof(old_log));
```
Why do we need to memset only during rename? Why do we need to memset at all?

pks-gitlab replied:
@knayak-gitlab The idea here is that we're passing ownership of the `old_log` over to the `updates` array. I'll restructure this a bit and add a comment to explain.

knayak-gitlab commented on src/libgit2/refdb_reftable.c:
```diff
@@ -0,0 +1,1868 @@
+	if ((error = refdb_reftable_stack_for(&stack, backend, REFDB_REFTABLE_STACK_MAIN)) < 0)
+		goto out;
+
+	if ((error = reftable_stack_compact_all(stack->stack, NULL)) < 0) {
```
So no way to do geometric compression?

pks-gitlab replied:
@knayak-gitlab I guess this can be added to the interfaces at a later point.

knayak-gitlab commented:
Some questions/comments from me. I mostly focused on the reftable code and not the infrastructure around it. Thanks.

pks-gitlab commented:
@knayak-gitlab Thanks for your review, back to you! :ping_pong:

knayak-gitlab commented:
> @knayak-gitlab Thanks for your review, back to you! 🏓

The changes look good now! Thanks.

jdidion commented:
I have a project that is blocked on this PR landing.

In short - users who initialize repos with `git init --ref-format=reftable` hit `Not a repository` on startup; we emit a friendly error but cannot open the repo via `git2::Repository::discover` until libgit2 supports reftable.

I am happy to test my project against a branch build if that would help. I am also willing to contribute an AI code review (Opus + GPT 5.5 + Gemini working as a “crew”) if that would be of help.

Thanks to @pks-gitlab for carrying this through multiple iterations.

ethomson commented:
Thanks @pks-gitlab for pushing on this!

ofek commented:
When might this be released?
